x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobConfigProvider.java

```diff
@@ -0,0 +1,396 @@
+                ElasticsearchMappings.DOC_TYPE, Job.documentId(jobId));
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, DeleteAction.INSTANCE, request, new ActionListener<DeleteResponse>() {
+            @Override
+            public void onResponse(DeleteResponse deleteResponse) {
+                if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                    actionListener.onFailure(ExceptionsHelper.missingJobException(jobId));
+                    return;
+                }
+
+                assert deleteResponse.getResult() == DocWriteResponse.Result.DELETED;
+                actionListener.onResponse(deleteResponse);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                actionListener.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * Get the job and update it by applying {@code jobUpdater} then index the changed job
+     * setting the version in the request. Applying the update may cause a validation error
+     * which is returned via {@code updatedJobListener}
+     *
+     * @param jobId The Id of the job to update
+     * @param update The job update
+     * @param maxModelMemoryLimit The maximum model memory allowed
+     * @param updatedJobListener Updated job listener
+     */
+    public void updateJob(String jobId, JobUpdate update, ByteSizeValue maxModelMemoryLimit, ActionListener<Job> updatedJobListener) {
+        GetRequest getRequest = new GetRequest(AnomalyDetectorsIndex.configIndexName(),
+                ElasticsearchMappings.DOC_TYPE, Job.documentId(jobId));
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, GetAction.INSTANCE, getRequest, new ActionListener<GetResponse>() {
+            @Override
+            public void onResponse(GetResponse getResponse) {
+                if (getResponse.isExists() == false) {
+                    updatedJobListener.onFailure(ExceptionsHelper.missingJobException(jobId));
+                    return;
+                }
+
+                long version = getResponse.getVersion();
+                BytesReference source = getResponse.getSourceAsBytesRef();
+                Job.Builder jobBuilder;
+                try {
+                     jobBuilder = parseJobLenientlyFromSource(source);
+                } catch (IOException e) {
+                    updatedJobListener.onFailure(new ElasticsearchParseException("failed to parse " + getResponse.getType(), e));
+                    return;
+                }
+
+                Job updatedJob;
+                try {
+                    // Applying the update may result in a validation error
+                    updatedJob = update.mergeWithJob(jobBuilder.build(), maxModelMemoryLimit);
+                } catch (Exception e) {
+                    updatedJobListener.onFailure(e);
+                    return;
+                }
+
+                try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
```

Comment:
and reuse `putJob` here? We'd have to factor out the index request in a private method that takes the `OpType` as a parameter but it would save us some code.